drivers/sensor/default_rtio_sensor.c

```diff
@@ -0,0 +1,371 @@
+	q31_t *q = (q31_t *)(buf + sizeof(struct sensor_data_generic_header) +
+			     num_channels * sizeof(struct sensor_data_generic_channel));
+
+	for (size_t i = 0; i < num_channels; ++i) {
+		struct sensor_value value[3];
+		int num_samples = SENSOR_CHANNEL_3_AXIS(channels[i]) ? 3 : 1;
+
+		/* Get the current channel requested by the user */
+		rc = sensor_channel_get(dev, channels[i], value);
+		if (rc != 0) {
+			LOG_ERR("Failed to get channel %d", channels[i]);
+			rtio_iodev_sqe_err(iodev_sqe, rc);
+			return;
+		}
+
+		/* Get the largest absolute value reading to set the scale for the channel */
+		uint32_t header_scale = 0;
+
+		for (int sample = 0; sample < num_samples; ++sample) {
+			uint32_t scale = (uint32_t)llabs((int64_t)value[sample].val1 + 1);
+
+			header_scale = MAX(header_scale, scale);
+		}
+		header->channel_info[i].channel = channels[i];
+		header->channel_info[i].shift = ilog2(header_scale - 1) + 1;
+
+		/*
+		 * Spread the q31 values. This is needed because some channels are 3D. If
+		 * the user specified one of those then num_samples will be 3; and we need to
+		 * produce 3 separate readings.
+		 */
+		for (int sample = 0; sample < num_samples; ++sample) {
+			/* Convert the value to micro-units */
+			int64_t value_u =
+				value[sample].val1 * INT64_C(1000000) + value[sample].val2;
+
+			/* Convert to q31 using the shift */
+			q[current_channel_index + sample] =
+				((value_u * ((INT64_C(1) << 31) - 1)) / 1000000) >>
+				header->channel_info[i].shift;
+
+			LOG_DBG("value[%d]=%s%d.%06d, q[%d]=%d", sample, value_u < 0 ? "-" : "",
+				abs((int)(value_u / 1000000)), abs((int)(value_u % 1000000)),
+				current_channel_index + sample, q[current_channel_index + sample]);
+		}
+		current_channel_index += num_samples;
+	}
+	rtio_iodev_sqe_ok(iodev_sqe, 0);
+}
+
+void z_sensor_processing_loop(struct rtio *ctx, sensor_processing_callback_t cb)
+{
+	void *userdata = NULL;
+	uint8_t *buf = NULL;
+	uint32_t buf_len = 0;
+	int rc;
+
+	/* Wait for a CQE */
+	struct rtio_cqe *cqe = rtio_cqe_consume_block(ctx);
+
+	/* Cache the data from the CQE */
+	rc = cqe->result;
+	userdata = cqe->userdata;
+	rtio_cqe_get_mempool_buffer(ctx, cqe, &buf, &buf_len);
+
+	/* Release the CQE */
+	rtio_cqe_release(ctx);
+
+	/* Call the callback */
+	cb(rc, buf, buf_len, userdata);
+
+	/* Release the memory */
+	rtio_release_buffer(ctx, buf, buf_len);
+}
+
+/**
+ * @brief Default decoder get frame count
+ *
+ * Default reader can only ever service a single frame at a time.
+ *
+ * @param[in]  buffer The data buffer to parse
+ * @param[out] frame_count The number of frames in the buffer (always 1)
+ * @return 0 in all cases
+ */
+static int get_frame_count(const uint8_t *buffer, uint16_t *frame_count)
+{
+	*frame_count = 1;
+	return 0;
+}
+
+/**
+ * @brief Default decoder get the timestamp of the first frame
+ *
+ * @param[in]  buffer The data buffer to parse
+ * @param[out] timestamp_ticks The timestamp of the first frame
```

Comment:
I'd be wary of using ticks here, I actually think the timestamp_ns is smarter as the clock source could be more precise than ticks.

Ticks to ns is possible, so nothing else would need to change but the signature of this for me to +1 today

Reply:
Done